### ISSUE TITLE
TOS fs

### ISSUE THREAD

**Issue by franklukas** [freemint/freemint#107]
Use a TOS/Windows partition made with hddriver does not work because you need the old TOS fs. The helmut kernel had it. It is possible to make a separately XFS driver or something else?

->   http://www.atari-forum.com/viewtopic.php?t=33823

**Comment by mikrosk** [comment]
Actually this isn't a OLDTOSFS issue, mzry used a 060 kernel which has it disabled.

**Comment by franklukas** [comment]
Can it be useful to make a "oldtosfs.xfs" kernel modul? I like to use media like CF cards partitioned with HDDriver as TOS/Windows compatible format.

**Comment by th-otto** [comment]
As mikrosk already said, that does not have to do anything with OLDTOSFS. That define is mainly meant for Hatari's GEMDOS Emulation.

**Comment by franklukas** [comment]
Why can I when I use such a media not open it under MiNT/XaAES with Teradesk as Desktop? Also there are errors at boottime.

**Comment by th-otto** [comment]
Can you upload an example image somewhere?

**Comment by franklukas** [comment]
I make the image with HDDriver. Not sure I make one with ByteSwap (make dialog) and one without? The CF card was made TOS and Windows with ByteSwap already.

In Thing and Teradesk comes up a error " pid 82 (thing): FATFS (J): vital data illegal (recsiz = 560, numcl = 15051)

->   https://www.dropbox.com/s/rvyyzeiearvqsc6/image.zip?dl=0

**Comment by th-otto** [comment]
The images are just images of the single partition; there is no partition table in it. I can mount the CF_32MB.IMG for example in ARAnyM without problems by assigning it to some PARTITION entry, and setting ByteSwap=yes

I guess when using such an image directly from a CF-Card, you need to create a partition table there first.

**Comment by mikrosk** [comment]
I would also recheck that it really works with Helmut's kernel out of the box.

**Comment by franklukas** [comment]
It works with helmut kernel.

**Comment by th-otto** [comment]
Normally the kernel tries to read a partition table from a device. Maybe Helmut added code to "fake" a partition table for a single partition, much like ARAnyM does. We'll check that.

**Comment by franklukas** [comment]
New Image of the device ...

https://www.dropbox.com/s/wkivhqtxwgzhqu4/32MB_CF.IMG.zip?dl=0

**Comment by th-otto** [comment]
Ok, that one looks different. It has a partition table with entry #0 having flag $40, and no type (i guess that is a dummy entry to make it compatible with windows), and entry #2 being a BGM partition with 63391 (0xf79f) sectors.

But again, i have no problem to access that image in ARAnyM by attaching it to IDE1 with ByteSwap=No:

![Screenshot_20200311_203818](https://user-images.githubusercontent.com/12425871/76457742-a4abd880-63d9-11ea-9722-e205fefe2819.png)

Only the filenames seem to have been truncated to their 8.3 format, instead of VFAT long names.

**Comment by franklukas** [comment]
I can use this 32MB CF under TOS and also under Windows or macOS. Its create with HDDriver as TOS/Windows with ByteSwap. This CF produce the error in post #7 under MiNT/XaAES ...

**Comment by th-otto** [comment]
In ARAnyM, i'm using the same current version of MinT&XaAES. Main difference though it that i'm not using HDDriver, but the builtin xhdi driver of EmuTOS:

```
EmuTOS version 0.9.12.1 20191107
unit 18 is managed by NatFeats: Empty
hda:  BGM
unit 19 is managed by NatFeats: DiskD
hdb:  MBR at 0 $06
```

(hda is an empty partition because the first IDE partition would otherwise be assigned to C:)

So maybe something is wrong with the byteswapping there, and HDDriver swaps it again?

**Comment by franklukas** [comment]
I use Atari TOS only, not EmuTOS ...

**Comment by th-otto** [comment]
Yes, all i want to say that there seems to be nothing wrong with mint.

**Comment by mikrosk** [comment]
@franklukas isn't possible that the HD Driver's mirror option behaves differently than mirroring the whole card using `dd` command (in Linux/Mac OS X) ?

That way we could be sure not only that we have the same "card" as you do but also we could get your boot setup (incl. the offending HD Driver).

**Comment by franklukas** [comment]
I make a image via dd under macOS ...

https://www.dropbox.com/s/t0xn4wu9tszmn6b/32mb.dd.zip?dl=0

**Comment by DavidGZ** [comment]
This is the commit that made that media to be unreadable with MiNT.
72c32cd5174a73515b0c3b73595a64db96968cf3

It works in Helmut's branch because it wasn't backported there.

**Comment by mikrosk** [comment]
Great find, David!

Related discussion:
- https://mikro.naprvyraz.sk/mint/freemint-list/Week-of-Mon-20160808/thread.html#74
- https://mikro.naprvyraz.sk/mint/freemint-list/Week-of-Mon-20161003/thread.html#123
- discussion (without resolution it would seem?): https://github.com/freemint/freemint/commit/4db32fddcfc95418009657ca6a9e4fc421953b31#commitcomment-22260709

**Comment by DavidGZ** [comment]
Thanks for the links.

The issue seems related with byte swapping, if I understood correctly the partition in the last image attached was created with HDDRIVER byte swapping enabled, I don't know why any XHDOSLimits() call to HDDRIVER changes the driver behaviour and it doesn't swap the bytes any more. And if byte swapping is enabled in Aranym for that image the partition can be read without issues.

**Comment by DavidGZ** [comment]
BTW in the limits set by the kernel we tell some lies, for the maximum cluster size in bytes we set 65536 and it should be 32768, block cache size marks this limit, as it has been discussed recently this is the reason of  the 2 GB limit for FAT16 partitions.

**Comment by DavidGZ** [comment]
@franklukas perhaps I'm missing something but I think that the problem could be in HDDRIVER.

**Comment by mikrosk** [comment]
That would be rather ironic because the whole XHDI commit was done per Uwe's request. He has his HD Driver forum: https://www.hddriver.net/forum where he reacts pretty quickly but I'm not sure how the bug report should look like?

**Comment by DavidGZ** [comment]
What it looks is happening is that HDDRIVER stops to swap bytes for disks partitioned with the "Byte Swapping" enable if receives a XHDOSLimits() XHDI call. I don't see why so I think it could be a bug in HDDRIVER.

**Comment by DavidGZ** [comment]
I'm reading now the thread you linked in his forum, there is information there about HDDRIVER doing a Mediach() after XHDOSLimits() call, I'll see if this could explain things but I don't think so because when we make Aranym swap bytes the disk readable again.

**Comment by mikrosk** [comment]
Ha, now I realised that Uwe has a github account. @uweseimet any chance you could take a look whether this is really happening?

**Comment by uweseimet** [comment]
I'm afraid I don't have any setup where I can test this with,
XHDOSLimits() will not stop HDDRIVER from swapping bytes. HDDRIVER indeed does not call Mediach() when XHDOSLimits() was called.
HDDRIVER essentially does this on a call to XHDOSLimits():

1. The new limits internally replace the old limits. (Usually, but not necessarily, the old limits are the original TOS limits, unless XHDOSLimits() is called several times. But this is usually not the case, as it is only called once by the software lifting those limits, e.g. Big-DOS.)
2. HDDRIVER updates _drvbits if the number of BIOS drives has been changed. Usually XHDOSLimits() increases the number of permitted drives. In case the call to XHDOSLimits() reduces the number of drives HDDRIVER will only clear the bits in _drvbits that are related to drives managed by HDDRIVER.
3. HDDRIVER *internally* marks a media change for all drives managed by HDDRIVER. As a result the next time a BPB is constructed the boot sector data will be re-evaluated based on the updated limits.
4. HDDRIVER calls XHDOSLimits() on the next driver in the XHDI chain, as per XHDI specification.
5. The old limit is returned as a result of the XHDOSLImits() call, as per XHDI specification.

This happens for each limit that is changed. XHDOSLimits() is not related to byte-swapping. Byte orders can only change when a medium is actually changed, and that's a different thing than changing the DOS limits.

**Comment by DavidGZ** [comment]
Hi @uweseimet, thank you for the info. I've spent some more time investigating this and I still think could be something wrong with HDDRIVER. I'm able to replicate what I described above on TOS too.

1. Boot TOS+Bigdos
2. Take a disk partitioned with TOS/Windows compatibility - byte swapping enable
3. Open it. Contents are OK
4. Run xhditest.ttp from Bigdos package
5. Open the disk again, contents are messed up

or

1. Boot TOS+Bigdos
2. Open a disk partitioned with TOS/Windows compatibility - byte swapping enable
3. It opens OK
4. Reboot
5. Run xhditest.ttp from Bigdos
5. Disk won't open

As it is said above it seems that for some reason HDDRIVER stops swapping bytes after it gets a XHDOSLimit() for setting any limit.

**Comment by uweseimet** [comment]
That's strange. Can you please check whether this also happens with purely TOS compatible media? Based on your results I will then try to reproduce this with a setup of my own.

**Comment by DavidGZ** [comment]
> That's strange. Can you please check whether this also happens with purely TOS compatible media? Based on your results I will then try to reproduce this with a setup of my own.

Yes, it does

> One more question: Does this also happen if you do _not_ run Big-DOS, but only XHDITEST? 
XHDITEST does not require Big-DOS to be installed.

Yes, it does.

> Please also confirm that there are no issues when the medium is not byte-swapped.

Yes, I confirm, if media is not byte-swapped everything works as expected

**Comment by uweseimet** [comment]
Thank you for running these tests. I noticed that you were not yet registered with the current version 11.01 of HDDRIVER. I've just updated your registration, so that you can download the current version now. Would be great if you could confirm that with HDDRIVER 11.01 you encounter the same issue with the simplest setup required, i.e. only TOS compatible, not Big-DOS, just running XHDITEST.

**Comment by DavidGZ** [comment]
I've just tested it with 11.01 and the issue is still present.

**Comment by uweseimet** [comment]
OK, I assume I will be able to reproduce this. Please send me your email address so that I can send you a beta version of HDDRIVER as soon as I have figured out what the issue is.

**Comment by uweseimet** [comment]
Just a quick update: After a lot of testing and tremendous support by David it looks as if I could track down and fix the problem. There were issues with how HDDRIVER maintained the byteswap state after operations that internally forced a media change.
Some final testing is remaining, but it's likely that HDDRIVER 11.02 will be released soon.

**Comment by mikrosk** [comment]
@DavidGZ and @uweseimet many thanks for your efforts!

**Comment by DavidGZ** [comment]
Fixed with HDDRIVER version 11.02
https://www.hddriver.net/en/news.html

Closing the issue
